### PR TITLE
Add -Wundef to --strict-warnings options

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Add -Wundef to --strict-warnings options.
This PR is against master

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
@richsalz Yes I did add the -Wundef to both, but I have only tested with gcc so far.

**Comment by richsalz** [comment]
i tested -wundef on clang.

**Comment by bernd-edlinger** [comment]
thanks!

File: crypto/ec/ec_asn1.c [hunk 1]
```diff
@@ -24,7 +24,7 @@ int EC_GROUP_get_basis_type(const EC_GROUP *group)
 
     /* Find the last non-zero element of group->poly[] */
     for (i = 0;
-         i < (int)OSSL_NELEM(group->poly) & group->poly[i] != 0;
+         i < (int)OSSL_NELEM(group->poly) && group->poly[i] != 0;
```
Comment (dot-asm) [inline crypto/ec/ec_asn1.c]: I'd argue that this should be separate commit. Rather insist...

Comment (richsalz) [inline crypto/ec/ec_asn1.c]: it will go away on rebase since @levitte aleady fixed it.

Comment (bernd-edlinger) [inline crypto/ec/ec_asn1.c]: I have also problems to add -Wundef on 1.0.2 because OSSL_NELEM is not defined there.

**Comment by richsalz** [comment]
See https://github.com/openssl/openssl/pull/2713

File: crypto/ec/ec_asn1.c (hunk 1 above)
Comment (dot-asm) [inline crypto/ec/ec_asn1.c]: > it will go away on rebase since @levitte aleady fixed it.

Still, unrelated changes to separate commits. Now this request would *have to* be re-based, normally one would have to wait out ci...

**Comment by bernd-edlinger** [comment]
Ok, then I assume I should do a rebase.

File: include/internal/refcount.h [hunk 2]
```diff
@@ -9,7 +9,8 @@
 #ifndef HEADER_INTERNAL_REFCOUNT_H
 # define HEADER_INTERNAL_REFCOUNT_H
 
-# if __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
+# if defined(__STDC_VERSION) && __STDC_VERSION__ >= 201112L \
```
Comment (bernd-edlinger) [inline include/internal/refcount.h]: Oops, I will fix that in a moment.

**Comment by richsalz** [comment]
i need the following added (and a rebase to catch the apps/rehash.c)
```
diff --git a/include/internal/refcount.h b/include/internal/refcount.h
index c8ebd5a..c70d95e 100644
--- a/include/internal/refcount.h
+++ b/include/internal/refcount.h
@@ -15,7 +15,8 @@
 # define HAVE_C11_ATOMICS
 # endif

-# if defined(HAVE_C11_ATOMICS) && ATOMIC_INT_LOCK_FREE > 0
+# if defined(HAVE_C11_ATOMICS) && defined(ATOMIC_INT_LOCK_FREE) \
+    && ATOMIC_INT_LOCK_FREE > 0

 # define HAVE_ATOMICS 1

```

**Comment by bernd-edlinger** [comment]
@richsalz 
> i need the following added (and a rebase to catch the apps/rehash.c)

done, and split up in smaller commits, as @dot-asm suggested.

**Comment by richsalz** [comment]
well since we already merged your PR for 1.1.0, the split is just an academic exercise :)
@dot-asm, please reconfirm and I will merge/squash.

**Comment by bernd-edlinger** [comment]
squashed again as an exercise ;)

**Comment by dot-asm** [comment]
Reconfirmed.

"Side note" means that no action is expected, this is likely something to consider for future.

**Comment by richsalz** [comment]
merged, thank you!
